lib/generators.nix

```diff
@@ -334,6 +334,78 @@ rec {
 ${expr "" v}
 </plist>'';
 
+  toLibconfig = { }:
```

Comment:
This can't currently generate valid config for the example. :(

Reply:
```git
diff --git a/lib/generators.nix b/lib/generators.nix
index 0eab0c076c1..23f57e5fdcd 100644
--- a/lib/generators.nix
+++ b/lib/generators.nix
@@ -393,6 +393,39 @@ ${expr "" v}
     v:
     let
       isFloat = builtins.isFloat or (x: false);
+      isHexString = str: builtins.match "^[+-]?0x[0-9a-fA-F]+$" str != null;
+      isOctalString = str: builtins.match "^[+-]?0[1-7][0-7]*$" str != null;
+      isLongString = str: builtins.match "^[+-]?[0-9]+L$" str != null;
+      isNumber = x: (isHexString x) || (isOctalString x) || (isLongString x);
+      isScalar = builtins.isString or isNumber or builtins.isFloat or builtins.isBool;
+
+      isArray = x: let
+        typed = builtins.groupBy type x;
+      in
+        !(typed ? "null") && !(typed ? "complex");
+
+      type = x:
+        with builtins;
+        if x == null then
+          "null"
+        else if isBool x then
+          "bool"
+        else if isInt x then
+          "int"
+        else if isFloat x then
+          "float"
+        else if isString x then
+          if isHexString x then
+            "hex"
+          else if isOctalString x then
+            "octal"
+          else if isLongString x then
+            "long"
+          else
+            "string"
+        else
+          "complex";
+
       expr = ind: x:
         with builtins;
         if x == null then
@@ -404,7 +437,10 @@ ${expr "" v}
         else if isFloat x then
           num ind x
         else if isString x then
-          str ind x
+          if isNumber x then
+            lit ind x
+          else
+            str ind x
         else if isPath x then
           path ind x
         else if isList x then
@@ -421,6 +457,7 @@ ${expr "" v}
       str = ind: x: literal ind ''"${x}"'';
       key = ind: x: literal ind "${x}: ";
       path = ind: x: literal ind (toString x);
+      lit = ind: x: literal ind "${x}";
 
       indent = ind: expr "\t${ind}";
 
@@ -430,11 +467,18 @@ ${expr "" v}
         '' (indent ind);
 
       list = ind: x:
-        libStr.concatStringsSep "\n" [
-          (literal ind "(")
-          (item ind x)
-          (literal ind ")")
-        ];
+        if isArray x then
+          libStr.concatStringsSep "\n" [
+            (literal ind "[")
+            (item ind x)
+            (literal ind "]")
+          ]
+        else
+          libStr.concatStringsSep "\n" [
+            (literal ind "(")
+            (item ind x)
+            (literal ind ")")
+          ];
 
       attrs = ind: x:
         libStr.concatStringsSep "\n" [
```

Here's a vaguely correct patch to allow the example to generate valid config. [Integers](https://hyperrealm.github.io/libconfig/libconfig_manual.html#Integer-Values), and [Longs](https://hyperrealm.github.io/libconfig/libconfig_manual.html#g_t64_002dbit-Integer-Values) need you to squint at strings to see if they are really number literals. (regex... uh... works). [Arrays](https://hyperrealm.github.io/libconfig/libconfig_manual.html#Arrays) are just ambiguous in nix config. The above "works". I'd be worried `(...)` where all elements had the same type is expected somewhere and the above would spit out `[...]` instead breaking some tool. Explaining some conventional naming strategy  (`key.array = [...]`) would work but be confusing to debug.

This seems like a place nix the language isn't suitable on it's own. Alternatives?

Reply:
I can't do stuff right now but I've been using this with [my Nix module](https://github.com/ckiee/nixfiles/blob/0ac57655f15500051736923287ca55f3c84b65e5/modules/logiops.nix#L14) and it's been working quite nicely for a while (excluding issues on wake from suspend occasionally). Here's the [`logid.conf`](https://i.ckie.dev/J0qoyzt.cfg).

> This can't currently generate valid config for the example. :(

It should be able to if you dare to manually convert hex to decimal (as seen in my link above)

> The libconfig implementation won't be good enough for users.

It's been good enough for me and that's why I'm here to share it. I'll apply your patch soon but I wrote just enough to have it work for me and I do not care about spec compliance as long as I'm not doing [anything surprising](https://en.wikipedia.org/wiki/Principle_of_least_astonishment). 

> This seems like a place nix the language isn't suitable on it's own. Alternatives?

Then I can't just do this:

```
ckie@cookiemonster ~/git/nixfiles -> c eval 'nodes.'$(hostname)'.config.services.logiops.settings.devices' | head -n5
[
  {
    "buttons": [
      {
        "action": {
```

Or override a value for a specific system.

Reply:
[@include](https://hyperrealm.github.io/libconfig/libconfig_manual.html#Include-Directives) should also be supportable with `"@include" = "...".

Reply:
I don't mean to imply what you've done isn't helpful. To be clear the example that I previously suggested will not work (`"cid": "0x56"`).

Regarding nix: this comment is too brief on my part. I'm doing parsing work with regex trying to apply types to strings. It works but likely could be better with different tools :shrug:. I also should not have implied I wanted you in particular to come up with an alternative.

Regarding surprise: I tried the least surprising thing to me :D. I produced an example using strings for hexadecimal values. Likely assuming `"..."` would be passed through without quotes as I wasn't aware of the types used by libconfig. Similarly I'm surprised `()` and `[]` are interchangeable in `keys:`. Poking at the spec I can guess why it would work.

Most of what I've attempted to address, and poorly communicated, is treating `lib.toLibconfig` as a workable tool for others. Your work around for the integer type is probably all that's needed for `logiops` and can be addressed in the `example`, `description`, and perhaps a comment on the generator.

Reply:
> I also should not have implied I wanted you in particular to come up with an alternative.

I think not sleeping for >15.5 hours probably had something to do with my rather harsh reply. Sorry :P

Anyway I applied your patch and looked at the spec and it's actually pretty small which makes me understand why you wanted it to be complaint. I guess I just don't like the "user" phrasing. Sorry again.